framework/util/mark_injected_commands.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+** Copyright (c) 2024 Valve Corporation
+** Copyright (c) 2024 LunarG, Inc.
+**
+** Permission is hereby granted, free of charge, to any person obtaining a
+** copy of this software and associated documentation files (the "Software"),
+** to deal in the Software without restriction, including without limitation
+** the rights to use, copy, modify, merge, publish, distribute, sublicense,
+** and/or sell copies of the Software, and to permit persons to whom the
+** Software is furnished to do so, subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in
+** all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#include "util/defines.h"
+#include "logging.h"
+#include "mark_injected_commands.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(util)
+
+static void BeginEndInjectedCommandsNoop() {}
+
+#ifdef GFXRECON_DEBUG_BUILD
+static bool injecting_api_calls_g = false;
+#endif
+
+static PFN_BeginInjectedCommands BeginInjectCommands_fp = BeginEndInjectedCommandsNoop;
+static PFN_EndInjectedCommands   EndInjectCommands_fp   = BeginEndInjectedCommandsNoop;
+
+extern "C" void SetInjectedCommandCallbacks(PFN_BeginInjectedCommands begin_fp, PFN_EndInjectedCommands end_fp)
+{
+    BeginInjectCommands_fp = begin_fp;
+    EndInjectCommands_fp   = end_fp;
+}
+
+void BeginInjectedCommands()
+{
+#ifdef GFXRECON_DEBUG_BUILD
+    if (injecting_api_calls_g)
+    {
+        GFXRECON_LOG_WARNING("Nested BeginInjectedCommands/EndInjectedCommands");
```

Comment:
Nice error check. My understanding of it is that if we hit this message it means that GFXR code is incorrectly using the Begin/End calls. If that is the case this could be upgraded to a `GFXRECON_ASSERT` since it indicates a code problem, not a runtime problem.